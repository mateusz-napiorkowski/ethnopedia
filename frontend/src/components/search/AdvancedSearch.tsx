import React, { useEffect, useState } from "react"
import { useFormik } from "formik"
import { useQuery } from "react-query"
import { getAllCategories } from "../../api/categories"
import { ReactComponent as PlusIcon } from "../../assets/icons/plus.svg"
import { ReactComponent as CloseIcon } from "../../assets/icons/close.svg"
import { ReactComponent as SearchLoopIcon } from "../../assets/icons/searchLoop.svg"
import LoadingPage from "../../pages/LoadingPage"
import { useLocation, useNavigate } from "react-router-dom"
import { v4 as uuidv4 } from "uuid"
import DOMPurify from "dompurify"
import Dropdown from "../Dropdown"

interface SearchComponentProps {
    collectionId: string
}

interface Rule {
    field: string
    value: string
    id: string
}

const AdvancedSearch: React.FC<SearchComponentProps> = ({ collectionId }) => {
    const navigate = useNavigate()
    const location = useLocation()
    const [rules, setRules] = useState<Rule[]>([])
    const [currentRuleCategory, setCurrentRuleCategory] = useState("")
    const [currentRuleValue, setCurrentRuleValue] = useState("")
    const [errorMessage, setErrorMessage] = useState<string | null>(null)
    const [duplicateCategoryError, setDuplicateCategoryError] = useState(false)
    const [valueError, setValueError] = useState(false)

    useEffect(() => {
        if (location.state && location.state.rules) {
            setRules(location.state.rules)
        } else {
            setRules([]) // set default if no rules
        }
    }, [location])

    const { data: categoriesData } = useQuery({
        queryKey: ["allCategories"],
        queryFn: () => getAllCategories(collectionId),
        enabled: !!collectionId,
    })

    const formik = useFormik({
        initialValues: {},
        onSubmit: () => {
            const queryString = rules.map(rule => `${rule.field}=${rule.value}`).join("&")
            navigate(`/collections/${collectionId}/artworks?${queryString}`, { state: { rules } })
        },
    })

    const handleCurrentRuleCategoryChange = (value: string) => {
        setCurrentRuleCategory(value)
        setDuplicateCategoryError(false) // resetujemy błąd, kiedy kategoria się zmienia
        setErrorMessage(null)
    }

    const handleCurrentRuleValueChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setErrorMessage(null)
        setCurrentRuleValue(event.target.value)
    }

    const handleAddRule = () => {
        const trimmedRawValue = currentRuleValue.trim()
        const sanitizedValue = DOMPurify.sanitize(trimmedRawValue)

        setValueError(false)
        setErrorMessage(null)
        setDuplicateCategoryError(false)

        if (!currentRuleCategory) {
            setErrorMessage("Wybierz kategorię przed dodaniem reguły.")
            setDuplicateCategoryError(true)
            return
        }

        // if (!trimmedRawValue) {
        //     setErrorMessage("Wprowadź wartość – to pole nie może być puste.")
        //     setValueError(true)
        //     return
        // }

        if (sanitizedValue !== trimmedRawValue) {
            setErrorMessage("Wartość zawiera niedozwolone znaki, np. <, >, lub inne specjalne znaki. Proszę usuń je i spróbuj ponownie.")
            setValueError(true)
            return
        }

        // TODO czy powinna być opcja dodania kilku regół dla tej samej kategorii? Łączenie na zasadzie OR?
        const isRuleCategoryUnique = rules.every(rule => currentRuleCategory !== rule.field)
        if (!isRuleCategoryUnique) {
            setErrorMessage("Reguła dla tej kategorii została już dodana. Wybierz inną kategorię.")
            setDuplicateCategoryError(true)
            return
        }

        setRules([...rules, {
            field: currentRuleCategory,
            value: sanitizedValue,
            id: uuidv4()
        }])

        setCurrentRuleCategory("")
        setCurrentRuleValue("")
        setErrorMessage(null)
        setValueError(false)
        setDuplicateCategoryError(false)
    }

    const handleDeleteRule = (id: string) => {
        setRules(rules.filter(rule => rule.id !== id))
    }

    if (!categoriesData) {
        return (
            <div data-testid="loading-advanced-search-container">
                <LoadingPage />
            </div>
        )
    }

    return (
        <div className="my-2" data-testid="advancedSearchComponent">
            <form onSubmit={formik.handleSubmit} className="flex flex-col gap-4">
                <div className="flex items-center gap-2 flex-wrap">
                    <Dropdown
                        options={categoriesData.categories.map((category: string) => ({
                            label: category,
                            value: category,
                        }))}
                        value={currentRuleCategory}
                        onChange={handleCurrentRuleCategoryChange}
                        placeholder="Wybierz kategorię"
                        error={duplicateCategoryError}
                    />

                    <input
                        name="value"
                        type="text"
                        maxLength={100}
                        onChange={handleCurrentRuleValueChange}
                        value={currentRuleValue}
                        className={`border p-2 rounded-lg ${valueError ? "border-red-500" : ""}`}
                    />

                    <button type="button" onClick={handleAddRule}
                            className="border-gray-800 flex items-center bg-gray-800 hover:bg-gray-700 text-white p-2 font-semibold"
                    >
                        <span className="mr-1">
                            <PlusIcon />
                        </span>
                        Dodaj regułę
                    </button>
                    <button type="submit" className="flex items-center font-semibold color-button p-2">
                        <span className="mr-1">
                            <SearchLoopIcon />
                        </span>
                        Wyszukaj
                    </button>
                    {errorMessage && (
                        <p className="w-full text-red-500 text-sm mt-2 ml-1">
                            {errorMessage}
                        </p>
                    )}
                </div>
            </form>

            <div data-testid="rules-container">
                {rules.map((rule) => (
                    <div key={rule.id} className="flex items-center gap-2 mt-4">
                        <button aria-label={`delete ${rule.field}`} onClick={() => handleDeleteRule(rule.id)} className="border-none p-0 mr-2">
                            <CloseIcon />
                        </button>
                        <span className="border border-blue-300 p-2 rounded-lg bg-blue-100 text-blue-500 font-semibold">
                            {rule.field}
                        </span>
                        <span className="border border-blue-300 p-2 rounded-lg bg-blue-100 text-blue-500 font-semibold">
                            {rule.value !== "" ? rule.value : "\u00A0"}
                        </span>
                    </div>
                ))}
            </div>
            <hr className="border-t border-gray-200 my-4 dark:border-gray-600" />
        </div>
    )
}

export default AdvancedSearch
