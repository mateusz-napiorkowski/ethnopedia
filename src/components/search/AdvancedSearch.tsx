const AdvancedSearch = () => {
    return <>
AdvancedSearch
    </>
}
export default AdvancedSearch