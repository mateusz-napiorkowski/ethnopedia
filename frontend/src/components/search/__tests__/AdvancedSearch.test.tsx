import '@testing-library/jest-dom';
import { render, waitFor } from '@testing-library/react'
import AdvancedSearch from "../AdvancedSearch"
import { QueryClient, QueryClientProvider } from "react-query";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import userEvent from "@testing-library/user-event";

const mockUseNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom') as any,
  useNavigate: () => mockUseNavigate,
}));

const mockGetAllCategories = jest.fn()
jest.mock('../../../api/categories', () => ({
    getAllCategories: () => mockGetAllCategories(),
}))

const queryClient = new QueryClient()
const user = userEvent.setup()

const renderPage = (
    queryClient: QueryClient,
    collection = "example collection" 
    ) => {
        return render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter initialEntries={[`/collections/${collection}/artworks`]}>
                    <Routes>
                        <Route path="/collections/:collection/artworks" element={<AdvancedSearch collectionName={collection} />}/>
                    </Routes>  
                </MemoryRouter>
            </QueryClientProvider>
        );
};

const fetchedCategories = {
    categories: [
        "Tytuł",
        "Tytuł.Podtytuł",
        "Artyści",
        "Rok"
    ]
}

describe("AdvancedSearch tests", () => {
    beforeEach(() => {
        jest.clearAllMocks();
        queryClient.clear();
    });

    it("should render loading state", () => {        
        const {getByTestId, queryByTestId} = renderPage(queryClient)

        expect(getByTestId("loading-advanced-search-container")).toBeInTheDocument()
        expect(queryByTestId("advancedSearchComponent")).not.toBeInTheDocument()
    })

    it("should render advanced search component after categoriesData is fetched from API", async () => {
        mockGetAllCategories.mockReturnValue(fetchedCategories)
        const {getByTestId, queryByTestId} = renderPage(queryClient)

        await waitFor(() => expect(getByTestId("advancedSearchComponent")).toBeInTheDocument())
        expect(queryByTestId("loading-advanced-search-container")).not.toBeInTheDocument()
    })

    it("should add and render search rules correctly", async () => {
        mockGetAllCategories.mockReturnValue(fetchedCategories)
        const {getByTestId, getByRole} = renderPage(queryClient)
        await waitFor(() => getByTestId("advancedSearchComponent"))
        const selectField = getByRole("combobox")
        const inputField = getByRole("textbox")
        const addRuleButton = getByRole("button", {
            name: /dodaj regułę/i
        })

        await user.selectOptions(selectField, "Tytuł")
        await user.type(inputField, "Przykładowy tytuł")
        await user.click(addRuleButton)

        await user.selectOptions(selectField, "Rok")
        await user.click(addRuleButton)

        await user.selectOptions(selectField, "Tytuł.Podtytuł")
        await user.type(inputField, "Przykładowy podtytuł")
        await user.click(addRuleButton)

        expect(getByTestId("rules-container")).toMatchSnapshot()
    })

    it("should clear out select and input elements after new rule is added", async () => {
        mockGetAllCategories.mockReturnValue(fetchedCategories)
        const {getByTestId, getByRole, getByText} = renderPage(queryClient)
        await waitFor(() => getByTestId("advancedSearchComponent"))
        const selectField = getByRole("combobox")
        const selectOption = (getByText("Wybierz kategorię") as HTMLOptionElement)
        const inputField = getByRole("textbox")
        const addRuleButton = getByRole("button", {
            name: /dodaj regułę/i
        })

        await user.selectOptions(selectField, "Tytuł")
        await user.type(inputField, "Przykładowy tytuł")
        await user.click(addRuleButton)

        expect(selectOption.selected).toBe(true)
        expect(inputField).toHaveValue("")
    })

    it("should delete appropriate search rules and render search rules correctly", async () => {
        mockGetAllCategories.mockReturnValue(fetchedCategories)
        const {getByTestId, getByRole, getByLabelText} = renderPage(queryClient)
        await waitFor(() => getByTestId("advancedSearchComponent"))
        const selectField = getByRole("combobox")
        const inputField = getByRole("textbox")
        const addRuleButton = getByRole("button", {
            name: /dodaj regułę/i
        })

        await user.selectOptions(selectField, "Tytuł")
        await user.type(inputField, "Przykładowy tytuł")
        await user.click(addRuleButton)

        await user.selectOptions(selectField, "Rok")
        await user.type(inputField, "1410")
        await user.click(addRuleButton)

        await user.click(getByLabelText("delete Tytuł"))

        await user.selectOptions(selectField, "Artyści")
        await user.type(inputField, "Jan Nowak")
        await user.click(addRuleButton)

        await user.selectOptions(selectField, "Tytuł.Podtytuł")
        await user.type(inputField, "Przykładowy podtytuł")
        await user.click(addRuleButton)

        await user.click(getByLabelText("delete Tytuł.Podtytuł"))

        expect(getByTestId("rules-container")).toMatchSnapshot()
    })

    it("should render element with error message when user tries to add rule with category name already present in another rule", async () => {
        mockGetAllCategories.mockReturnValue(fetchedCategories)
        const {getByTestId, getByRole, getByText} = renderPage(queryClient)
        await waitFor(() => getByTestId("advancedSearchComponent"))
        const selectField = getByRole("combobox")
        const inputField = getByRole("textbox")
        const addRuleButton = getByRole("button", {
            name: /dodaj regułę/i
        })

        await user.selectOptions(selectField, "Tytuł")
        await user.type(inputField, "Przykładowy tytuł")
        await user.click(addRuleButton)

        await user.selectOptions(selectField, "Tytuł")
        await user.type(inputField, "Przykładowy tytuł 2")
        await user.click(addRuleButton)

        expect(getByText(/reguła z tą nazwą kategorii już istnieje./i)).toBeInTheDocument()
    })

    it("should not render element with error message after user selects any option from select element", async () => {
        mockGetAllCategories.mockReturnValue(fetchedCategories)
        const {getByTestId, getByRole, queryByText} = renderPage(queryClient)
        await waitFor(() => getByTestId("advancedSearchComponent"))
        const selectField = getByRole("combobox")
        const inputField = getByRole("textbox")
        const addRuleButton = getByRole("button", {
            name: /dodaj regułę/i
        })

        await user.selectOptions(selectField, "Tytuł")
        await user.type(inputField, "Przykładowy tytuł")
        await user.click(addRuleButton)

        await user.selectOptions(selectField, "Tytuł")
        await user.type(inputField, "Przykładowy tytuł 2")
        await user.click(addRuleButton)

        await user.selectOptions(selectField, "Rok")

        expect(queryByText(/reguła z tą nazwą kategorii już istnieje./i)).not.toBeInTheDocument()
    })

    it("should navigate to the same page with correctly altered query string and component state after search button is clicked", async () => {
        mockGetAllCategories.mockReturnValue(fetchedCategories)
        const {getByTestId, getByRole } = renderPage(queryClient)
        await waitFor(() => getByTestId("advancedSearchComponent"))
        const selectField = getByRole("combobox")
        const inputField = getByRole("textbox")
        const addRuleButton = getByRole("button", {
            name: /dodaj regułę/i
        })
        const searchButton = getByRole("button", {
            name: /wyszukaj/i
        })

        await user.selectOptions(selectField, "Tytuł")
        await user.type(inputField, "tytuł")
        await user.click(addRuleButton)
        
        await user.selectOptions(selectField, "Tytuł.Podtytuł")
        await user.type(inputField, "podtytuł")
        await user.click(addRuleButton)

        await user.click(searchButton)

        expect(mockUseNavigate).toHaveBeenCalledWith(
            `/collections/example collection/artworks?Tytuł=tytuł&Tytuł.Podtytuł=podtytuł`,
            {"state":
                {"rules": [
                    {"field": "Tytuł", "id": expect.any(String), "value": "tytuł"},
                    {"field": "Tytuł.Podtytuł", "id": expect.any(String), "value": "podtytuł"}
                ]}
            }
        )
    })
})