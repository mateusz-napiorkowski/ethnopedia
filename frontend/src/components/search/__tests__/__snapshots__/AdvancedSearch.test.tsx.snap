// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvancedSearch tests should add and render search rules correctly 1`] = `
<div
  data-testid="rules-container"
>
  <div
    class="flex items-center gap-2 mt-4"
  >
    <button
      aria-label="delete Tytuł"
      class="border-none p-0 mr-2"
    >
      <svg />
    </button>
    <span
      class="border border-blue-300 p-2 rounded-lg bg-blue-100 text-blue-500 font-semibold"
    >
      Tytuł
    </span>
    <span
      class="border border-blue-300 p-2 rounded-lg bg-blue-100 text-blue-500 font-semibold"
    >
      Przykładowy tytuł
    </span>
  </div>
  <div
    class="flex items-center gap-2 mt-4"
  >
    <button
      aria-label="delete Rok"
      class="border-none p-0 mr-2"
    >
      <svg />
    </button>
    <span
      class="border border-blue-300 p-2 rounded-lg bg-blue-100 text-blue-500 font-semibold"
    >
      Rok
    </span>
    <span
      class="border border-blue-300 p-2 rounded-lg bg-blue-100 text-blue-500 font-semibold"
    >
       
    </span>
  </div>
  <div
    class="flex items-center gap-2 mt-4"
  >
    <button
      aria-label="delete Tytuł.Podtytuł"
      class="border-none p-0 mr-2"
    >
      <svg />
    </button>
    <span
      class="border border-blue-300 p-2 rounded-lg bg-blue-100 text-blue-500 font-semibold"
    >
      Tytuł.Podtytuł
    </span>
    <span
      class="border border-blue-300 p-2 rounded-lg bg-blue-100 text-blue-500 font-semibold"
    >
      Przykładowy podtytuł
    </span>
  </div>
</div>
`;

exports[`AdvancedSearch tests should delete appropriate search rules and render search rules correctly 1`] = `
<div
  data-testid="rules-container"
>
  <div
    class="flex items-center gap-2 mt-4"
  >
    <button
      aria-label="delete Rok"
      class="border-none p-0 mr-2"
    >
      <svg />
    </button>
    <span
      class="border border-blue-300 p-2 rounded-lg bg-blue-100 text-blue-500 font-semibold"
    >
      Rok
    </span>
    <span
      class="border border-blue-300 p-2 rounded-lg bg-blue-100 text-blue-500 font-semibold"
    >
      1410
    </span>
  </div>
  <div
    class="flex items-center gap-2 mt-4"
  >
    <button
      aria-label="delete Artyści"
      class="border-none p-0 mr-2"
    >
      <svg />
    </button>
    <span
      class="border border-blue-300 p-2 rounded-lg bg-blue-100 text-blue-500 font-semibold"
    >
      Artyści
    </span>
    <span
      class="border border-blue-300 p-2 rounded-lg bg-blue-100 text-blue-500 font-semibold"
    >
      Jan Nowak
    </span>
  </div>
</div>
`;
